// @economist/design-system/src/common/components/footer/footer-list.js
import React from 'react';
import PropTypes from 'prop-types';
import NavigationLink from '../navigation-link/navigation-link';
import css from './footer.css';

const FooterList = ({ links, secondary, major, minor, inverse }) => (
  <ul className={css['ds-footer__link-list']}>
    {links.map(({ url, id, name, amp, ...otherProps }) => (
      <li key={url} id={id}>
        <NavigationLink
          href={url}
          text={name}
          secondary={secondary}
          major={major}
          minor={minor}
          inverse={inverse}
          {...otherProps}
        />
      </li>
    ))}
  </ul>
);

FooterList.defaultProps = {
  secondary: false,
  major: false,
  minor: false,
  inverse: false,
};

FooterList.propTypes = {
  links: PropTypes.arrayOf(
    PropTypes.shape({
      url: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      id: PropTypes.string,
    }),
  ).isRequired,
  secondary: PropTypes.bool,
  major: PropTypes.bool,
  minor: PropTypes.bool,
  inverse: PropTypes.bool,
};

export default FooterList;
