// @economist/design-system/src/common/components/assets/assets.stories.js
import React from 'react';
import {
  Avatar,
  AvatarPlaceholder,
  BrandEspresso,
  BrandEspressoWeb,
  GraphicAudioChicago,
  GraphicAudioLondon,
  GraphicNewsletterChicago,
  GroupLogo,
  GroupLogoMasthead,
  GroupWordmark,
  IconCheck,
  IconCopy,
  ApplePodcasts,
  GooglePodcasts,
  Podcasts,
  Logo,
  LogoKeyline,
  MonogramCircle,
  MonogramRounded,
  MonogramSquare,
  SocialIconFacebook,
  SocialIconInstagram,
  SocialIconLinkedIn,
  SocialIconRss,
  SocialIconTwitter,
  SocialIconWhatsApp,
  SocialIconYoutube,
  Wordmark,
  WordmarkStacked,
  Wordmark1843,
  Wordmark1843Plain,
  WorldAheadWordmark,
} from './assets';

export default {
  title: 'common/Assets',
};

export const avatar = () => <Avatar />;
export const avatarPlaceholder = () => <AvatarPlaceholder />;
export const brandEspresso = () => <BrandEspresso />;
export const brandEspressoWeb = () => <BrandEspressoWeb />;
export const graphicAudioChicago = () => <GraphicAudioChicago />;
export const graphicAudioLondon = () => <GraphicAudioLondon />;
export const graphicNewsletterChicago = () => <GraphicNewsletterChicago />;
export const groupLogo = () => <GroupLogo />;
export const groupLogoMasthead = () => <GroupLogoMasthead />;
export const groupWordmark = () => <GroupWordmark />;
export const iconCheck = () => <IconCheck />;
export const iconCopy = () => <IconCopy />;
export const applePodcasts = () => <ApplePodcasts />;
export const googlePodcasts = () => <GooglePodcasts />;
export const podcasts = () => <Podcasts />;
export const logo = () => <Logo />;
export const logoKeyline = () => <LogoKeyline />;
export const monogramCircle = () => <MonogramCircle />;
export const monogramRounded = () => <MonogramRounded />;
export const monogramSquare = () => <MonogramSquare />;
export const socialIconFacebook = () => <SocialIconFacebook />;
export const socialIconInstagram = () => <SocialIconInstagram />;
export const socialIconLinkedIn = () => <SocialIconLinkedIn />;
export const socialIconRss = () => <SocialIconRss />;
export const socialIconTwitter = () => <SocialIconTwitter />;
export const socialIconWhatsApp = () => <SocialIconWhatsApp />;
export const socialIconYoutube = () => <SocialIconYoutube />;
export const wordmark = () => <Wordmark />;
export const wordmarkStacked = () => <WordmarkStacked />;
export const wordmark1843 = () => <Wordmark1843 />;
export const wordmark1843Plain = () => <Wordmark1843Plain />;
export const worldAheadWordmark = () => <WorldAheadWordmark />;
