// @economist/design-system/src/common/components/assets/assets.js
import Avatar from './avatar/avatar';
import BrandEspresso from './brands/espresso';
import BrandEspressoWeb from './brands/espresso-web';
import AvatarPlaceholder from './avatar/avatar-placeholder';
import GraphicAudioChicago from './graphic/audio-chicago';
import GraphicAudioLondon from './graphic/audio-london';
import GraphicNewsletterChicago from './graphic/newsletter-chicago';
import GroupLogo from './group-logo/group-logo';
import GroupLogoMasthead from './group-logo/group-logo-masthead';
import GroupWordmark from './group-wordmark/group-wordmark';
import IconCheck from './icon/icon-check';
import IconCopy from './icon/icon-copy';
import ApplePodcasts from './icon/apple-podcasts';
import GooglePodcasts from './icon/google-podcats';
import Podcasts from './icon/podcasts';
import Logo from './logo/logo';
import LogoKeyline from './logo/logo-keyline';
import MonogramCircle from './monogram/monogram-circle';
import MonogramRounded from './monogram/monogram-rounded';
import MonogramSquare from './monogram/monogram-square';
import SocialIconFacebook from './social-icons/facebook';
import SocialIconInstagram from './social-icons/instagram';
import SocialIconLinkedIn from './social-icons/linked-in';
import SocialIconRss from './social-icons/rss';
import SocialIconTwitter from './social-icons/twitter';
import SocialIconWhatsApp from './social-icons/whatsapp';
import SocialIconYoutube from './social-icons/youtube';
import Wordmark from './wordmark/wordmark';
import WordmarkStacked from './wordmark/wordmark-stacked';
import Wordmark1843 from './brands/1843-wordmark-magazine';
import Wordmark1843Plain from './brands/1843-wordmark-plain';
import WorldAheadWordmark from './brands/the-world-ahead-wordmark';

export {
  Avatar,
  AvatarPlaceholder,
  BrandEspresso,
  BrandEspressoWeb,
  GraphicAudioChicago,
  GraphicAudioLondon,
  GraphicNewsletterChicago,
  GroupLogo,
  GroupLogoMasthead,
  GroupWordmark,
  IconCheck,
  IconCopy,
  ApplePodcasts,
  GooglePodcasts,
  Podcasts,
  Logo,
  LogoKeyline,
  MonogramCircle,
  MonogramRounded,
  MonogramSquare,
  SocialIconFacebook,
  SocialIconInstagram,
  SocialIconLinkedIn,
  SocialIconRss,
  SocialIconTwitter,
  SocialIconWhatsApp,
  SocialIconYoutube,
  Wordmark,
  WordmarkStacked,
  Wordmark1843,
  Wordmark1843Plain,
  WorldAheadWordmark,
};
