// @economist/design-system/src/common/components/assets/idents-columns/index.js
import BagehotsNotebook from './bagehots-notebook';
import BartlebysNotebook from './bartlebys-notebook';
import ButtonwoodsNotebook from './buttonwoods-notebook';
import CharlemagnesNotebook from './charlemagnes-notebook';
import DemocracyInAmerica from './democracy-in-america';
import Erasmus from './erasmus';
import FreeExchange from './free-exchange';
import GameTheory from './game-theory';
import GraphicDetail from './graphic-detail';
import Gulliver from './gulliver';
import Kaffeeklatsch from './kaffeeklatsch';
import OpenFuture2018 from './open-future-2018';
import OpenFuture2019 from './open-future-2019';
import Prospero from './prospero';
import TheEconomistExplains from './the-economist-explains';

export const IDENTS_COLUMNS = {
  'bagehots-notebook': BagehotsNotebook,
  'bartlebys-notebook': BartlebysNotebook,
  'buttonwoods-notebook': ButtonwoodsNotebook,
  'charlemagnes-notebook': CharlemagnesNotebook,
  'democracy-in-america': DemocracyInAmerica,
  erasmus: Erasmus,
  'free-exchange': FreeExchange,
  'game-theory': GameTheory,
  'graphic-detail': GraphicDetail,
  gulliver: Gulliver,
  kaffeeklatsch: Kaffeeklatsch,
  'open-future-2018': OpenFuture2018,
  'open-future-2019': OpenFuture2019,
  prospero: Prospero,
  'the-economist-explains': TheEconomistExplains,
};

export const IDENT_NAMES = Object.keys(IDENTS_COLUMNS).filter(
  name => name !== '__filemeta',
  // Fixes issue with docz exporting an additional property: doczjs/docz#875"
);

export default IDENTS_COLUMNS;
