// @economist/design-system/src/common/components/masthead/account-login.js
import React from 'react';
import PropTypes from 'prop-types';
import NavigationLink from '../navigation-link/navigation-link';
import AccountDisclosure from './account-disclosure';

const AccountLogin = ({
  isLoggedIn,
  loginUrl,
  logoutUrl,
  inverse,
  toggleAccount,
  expanded,
  className,
  fullName,
}) => (
  <>
    {isLoggedIn && !logoutUrl && (
      <AccountDisclosure
        onClick={toggleAccount}
        text='My Economist'
        inverse={inverse}
        expanded={expanded}
        className={className}
        fullName={fullName}
      />
    )}
    {isLoggedIn && logoutUrl && (
      <NavigationLink href={logoutUrl} text='Log out' inverse={inverse} />
    )}
    {!isLoggedIn && (
      <NavigationLink href={loginUrl} text='Log in' inverse={inverse} />
    )}
  </>
);
AccountLogin.defaultProps = {
  isLoggedIn: false,
  loginUrl: '',
  logoutUrl: '',
  inverse: false,
  toggleAccount: () => {},
  expanded: false,
  className: '',
  fullName: '',
};

AccountLogin.propTypes = {
  isLoggedIn: PropTypes.bool,
  loginUrl: PropTypes.string,
  logoutUrl: PropTypes.string,
  inverse: PropTypes.bool,
  toggleAccount: PropTypes.func,
  expanded: PropTypes.bool,
  className: PropTypes.string,
  fullName: PropTypes.string,
};

export default AccountLogin;
