// @economist/design-system/src/common/components/masthead/masthead-secondary-navigation.js
import React from 'react';
import PropTypes from 'prop-types';
import NavigationList from './navigation-list';
import NavigationLink from '../navigation-link/navigation-link';
import MastheadSecondaryLinks from './masthead-secondary-links';
import 'reset-css/reset.css';
import '../grid/grid';
import '../media-queries/media-queries';
import css from './masthead.css';

const accountLogin = (
  isLoggedIn,
  accountData,
  menuIsOpen,
  username,
  showNavToggle,
  showAccountNav,
  logoutUrl,
) => (
  <>
    {isLoggedIn && !showAccountNav && (
      <li
        className={`${css['ds-masthead-nav-beta__item']} ${
          css['ds-masthead-nav-beta__item--log-in']
        }`}
      >
        <NavigationLink href={logoutUrl} text='Log out' inverse />
      </li>
    )}

    {isLoggedIn && showAccountNav && (
      <NavigationList
        type='account-logged-in'
        accountData={accountData}
        menuIsOpen={menuIsOpen}
        showNavToggle={showNavToggle}
        username={username}
      />
    )}
    {!isLoggedIn && (
      <NavigationList
        type='account-logged-out'
        accountData={accountData}
        menuIsOpen={menuIsOpen}
        showNavToggle={showNavToggle}
      />
    )}
  </>
);

const MastheadSecondaryNavigation = ({
  links,
  menuIsOpen,
  showLogin,
  showNavToggle,
  accountData,
  username,
  isLoggedIn,
  showAccountNav,
  loginUrl,
  logoutUrl,
}) => {
  const accountLoginLinks = accountLogin(
    isLoggedIn,
    accountData,
    menuIsOpen,
    username,
    showNavToggle,
    showAccountNav,
    logoutUrl,
  );

  return (
    <nav className={css['ds-masthead-nav-lists']}>
      <ul>
        {!isLoggedIn && loginUrl && (
          <li
            className={`${css['ds-masthead-nav-beta__item']} ${
              css['ds-masthead-nav-beta__item--log-in']
            }`}
          >
            <NavigationLink href={loginUrl} text='Log in' inverse />
          </li>
        )}
        {showLogin && isLoggedIn && !showAccountNav && accountLoginLinks}
        {links && <MastheadSecondaryLinks links={links} inverse />}
        {showLogin && isLoggedIn && showAccountNav && accountLoginLinks}
      </ul>
    </nav>
  );
};

MastheadSecondaryNavigation.defaultProps = {
  accountData: null,
  isLoggedIn: false,
  links: null,
  loginUrl: '',
  logoutUrl: '',
  showLogin: false,
  showNavToggle: false,
  showAccountNav: true,
  username: '',
};

MastheadSecondaryNavigation.propTypes = {
  accountData: PropTypes.objectOf(PropTypes.array),
  isLoggedIn: PropTypes.bool,
  links: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string,
      href: PropTypes.string,
      id: PropTypes.string,
    }),
  ),
  loginUrl: PropTypes.string,
  logoutUrl: PropTypes.string,
  menuIsOpen: PropTypes.bool.isRequired,
  showLogin: PropTypes.bool,
  showNavToggle: PropTypes.bool,
  showAccountNav: PropTypes.bool,
  username: PropTypes.string,
};

export default MastheadSecondaryNavigation;
