// @economist/design-system/src/common/components/masthead/masthead-links.js
import React from 'react';
import PropTypes from 'prop-types';
import NavigationLink from '../navigation-link/navigation-link';
import { urls } from '../../constants';
import 'reset-css/reset.css';
import '../color/color';
import '../grid/grid';
import '../interactions/interactions';
import '../media-queries/media-queries';
import '../typography/typography';
import css from './masthead.css';

const MastheadLinks = ({ inverse, selectedLink, links }) => (
  <>
    <li
      className={`${css['ds-masthead-nav-alpha__item']} ${
        css['ds-masthead-nav-alpha__item--weekly-edition']
      }`}
    >
      <NavigationLink
        text='Weekly edition'
        href={links.weeklyEdition}
        inverse={inverse}
        selected={selectedLink === 'weekly-edition'}
        className='weekly-edition-link'
      />
    </li>
    <li
      className={`${css['ds-masthead-nav-alpha__item']} ${
        css['ds-masthead-nav-alpha__item--the-world-in-brief']
      }`}
    >
      <NavigationLink
        text='The world in brief'
        href={links.theWorldInBrief}
        inverse={inverse}
        selected={selectedLink === 'the-world-in-brief'}
        className='the-world-in-brief-link'
      />
    </li>
  </>
);

MastheadLinks.defaultProps = {
  inverse: false,
  selectedLink: null,
  links: urls,
};

MastheadLinks.propTypes = {
  inverse: PropTypes.bool,
  selectedLink: PropTypes.oneOf([
    'weekly-edition',
    'the-world-in-brief',
    'today',
  ]),
  links: PropTypes.shape({
    economist: PropTypes.string.isRequired,
    weeklyEdition: PropTypes.string.isRequired,
    theWorldInBrief: PropTypes.string.isRequired,
  }),
};

export default MastheadLinks;
