// @economist/design-system/src/common/components/masthead/masthead-secondary-links.js
import React from 'react';
import PropTypes from 'prop-types';
import NavigationLink from '../navigation-link/navigation-link';
import 'reset-css/reset.css';
import '../color/color';
import '../grid/grid';
import '../interactions/interactions';
import '../media-queries/media-queries';
import '../typography/typography';
import css from './masthead.css';

const MastheadSecondaryLinks = ({ links, inverse, selectedLink }) => (
  <>
    {links.map(({ text, href, id }, i) => (
      <li key={id} className={css['ds-masthead-nav-alpha__item']}>
        <NavigationLink
          text={text}
          href={href}
          inverse={inverse}
          selected={i === selectedLink}
        />
      </li>
    ))}
  </>
);

MastheadSecondaryLinks.defaultProps = {
  inverse: false,
  selectedLink: null,
};

MastheadSecondaryLinks.propTypes = {
  inverse: PropTypes.bool,
  links: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string,
      href: PropTypes.string,
      id: PropTypes.string,
    }),
  ).isRequired,
  selectedLink: PropTypes.number,
};

export default MastheadSecondaryLinks;
