// @economist/design-system/src/common/components/masthead/navigation-list.js
import React from 'react';
import PropTypes from 'prop-types';
import NavigationLink from '../navigation-link/navigation-link';
import NavigationControl from './navigation-control';
import { urls, getNavigationListData } from '../../constants';
import 'reset-css/reset.css';
import '../color/color';
import '../rule/rule';
import '../typography/typography';
import css from './navigation-list.css';

const renderListItem = item => {
  return (
    <li key={item.name.toString()}>
      <NavigationLink inverse href={item.url} text={item.name} />
    </li>
  );
};

const NavigationList = ({
  type,
  baseUrl,
  menuIsOpen,
  showNavToggle,
  username,
  accountData,
  listIsOpen,
}) => {
  let navigationListComponent;
  let navigationListName;
  const navigationListData = getNavigationListData(baseUrl);

  switch (type) {
    case 'account-logged-in':
      navigationListComponent = accountData[type].map(item =>
        renderListItem(item),
      );
      if (username) {
        navigationListComponent.unshift(
          <li key='username'>
            <strong className={css['ds-navigation-list__header-username']}>
              {username}
            </strong>
          </li>,
        );
      }
      navigationListName = 'My account';
      break;

    case 'feature':
      navigationListComponent = navigationListData[type].map(item =>
        renderListItem(item),
      );
      navigationListName = 'Featured';
      break;
    case 'more':
      navigationListComponent = navigationListData[type].map(item =>
        renderListItem(item),
      );
      navigationListName = 'More';
      break;
    case 'section':
      navigationListComponent = navigationListData[type].map(item =>
        renderListItem(item),
      );
      navigationListName = 'Sections';
      break;
    default:
      return null;
  }
  return (
    <NavigationControl
      title={navigationListName}
      type={type}
      menuIsOpen={menuIsOpen}
      showNavToggle={showNavToggle}
      listIsOpen={listIsOpen}
    >
      {navigationListComponent}
    </NavigationControl>
  );
};

NavigationList.defaultProps = {
  baseUrl: urls.economist,
  accountData: {},
  username: '',
  listIsOpen: false,
};

NavigationList.propTypes = {
  baseUrl: PropTypes.string,
  menuIsOpen: PropTypes.bool.isRequired,
  accountData: PropTypes.objectOf(PropTypes.array),
  showNavToggle: PropTypes.bool.isRequired,
  type: PropTypes.oneOf([
    'blog',
    'account-logged-in',
    'account-logged-out',
    'more',
    'section',
    'feature',
  ]).isRequired,
  username: PropTypes.string,
  listIsOpen: PropTypes.bool,
};

export default NavigationList;
