// @economist/design-system/src/common/constants.js
/* eslint-disable import/prefer-default-export */
import { Element } from 'react';
import PropTypes from 'prop-types';

export const breakpointSmall = 380;
export const breakpointMedium = 600;
export const breakpointLarge = 960;
export const breakpointXLarge = 1088;

export const urls = {
  economist: 'https://www.economist.com',
  weeklyEdition: 'https://www.economist.com/weeklyedition',
  search: 'https://www.economist.com/search',
  theWorldInBrief: 'https://www.economist.com/the-world-in-brief',
  subscribe: 'https://subscription.economist.com/DE/EngCore/Ecom/Masthead',
  subscribeFooter: 'https://www.economist.com/subscribe',
  myEconomist: 'https://my.economist.com',
  cookiePolicy:
    'https://www.economistgroup.com/results_and_governance/governance/Cookies_information.html',
  cookieManage: 'https://l3.evidon.com/site/2254/668/6?lang=en-gb',
};

/**
 * URLs can be overwritten as props in some components.
 * The object should only accept valid keys, as defined in urls
 */
export const urlsPropShape = PropTypes.shape(
  Object.keys(urls).reduce(
    (shape, key) => ({
      ...shape,
      [key]: PropTypes.string,
    }),
    {},
  ),
);

export const accountData = {
  'account-logged-in': [
    {
      name: 'My Economist',
      url: '/my-economist',
    },
    {
      name: 'Saved stories',
      url: '/saved-stories',
    },
    {
      name: 'Admin account',
      url: '',
    },
    {
      name: 'My subscription',
      url: '',
    },
    {
      name: 'Personal information',
      url: '',
    },
    {
      name: 'Billing',
      url: '',
    },
    {
      name: 'Newsletters & preferences',
      url: '',
    },
    {
      name: 'Gift subscriptions',
      url: '',
    },
    {
      name: 'Help',
      url: '',
    },
    {
      name: 'Log out',
      url: '',
    },
  ],
  'account-logged-in-main': [
    {
      name: 'Saved stories',
      url: '/saved-stories',
    },
    {
      name: 'Admin account',
      url: '',
    },
    {
      name: 'Account',
      url: '/my-economist',
    },
    {
      name: 'Log out',
      url: '',
    },
  ],
  'account-logged-out': [
    {
      name: 'Log in',
      url: '',
    },
  ],
};

export const clientAdminData = {
  name: 'My group subscription',
  url: '',
};

// We temporarily store navigation list data here, eventually
// this should be provided by the content platform.
export const getNavigationListData = (baseUrl = urls.economist) => ({
  feature: [
    {
      name: 'War in Ukraine',
      url: `${baseUrl}/ukraine-crisis`,
    },
    {
      name: 'Recession watch',
      url: `${baseUrl}/recession`,
    },
    {
      name: 'The World Ahead 2023',
      url: `${baseUrl}/the-world-ahead-2023`,
    },
    {
      name: 'US politics',
      url: `${baseUrl}/president-joe-biden`,
    },
    {
      name: 'Climate change',
      url: `${baseUrl}/climate-change`,
    },
    {
      name: 'Coronavirus',
      url: `${baseUrl}/coronavirus-pandemic`,
    },
    {
      name: '1843 magazine',
      url: `${baseUrl}/1843` /* CP response will fail with trailing slash */,
    },
    {
      name: 'The world in brief',
      url: `${baseUrl}/the-world-in-brief`,
    },
  ],
  section: [
    {
      name: 'The world this week',
      url: `${baseUrl}/the-world-this-week`,
    },
    {
      name: 'Leaders',
      url: `${baseUrl}/leaders`,
    },
    {
      name: 'Letters',
      url: `${baseUrl}/letters`,
    },
    {
      name: 'Briefing',
      url: `${baseUrl}/briefing`,
    },
    {
      name: 'United States',
      url: `${baseUrl}/united-states`,
    },
    {
      name: 'The Americas',
      url: `${baseUrl}/the-americas`,
    },
    {
      name: 'Asia',
      url: `${baseUrl}/asia`,
    },
    {
      name: 'China',
      url: `${baseUrl}/china`,
    },
    {
      name: 'Middle East & Africa',
      url: `${baseUrl}/middle-east-and-africa`,
    },
    {
      name: 'Europe',
      url: `${baseUrl}/europe`,
    },
    {
      name: 'Britain',
      url: `${baseUrl}/britain`,
    },
    {
      name: 'International',
      url: `${baseUrl}/international`,
    },
    {
      name: 'Business',
      url: `${baseUrl}/business`,
    },
    {
      name: 'Finance & economics',
      url: `${baseUrl}/finance-and-economics`,
    },
    {
      name: 'Science & technology',
      url: `${baseUrl}/science-and-technology`,
    },
    {
      name: 'Culture',
      url: `${baseUrl}/culture`,
    },
    {
      name: 'Graphic detail',
      url: `${baseUrl}/graphic-detail`,
    },
    {
      name: 'Obituary',
      url: `${baseUrl}/obituary`,
    },
    {
      name: 'Special reports',
      url: `${baseUrl}/special-reports`,
    },
    {
      name: 'Technology Quarterly',
      url: `${baseUrl}/technology-quarterly`,
    },
    {
      name: 'Essay',
      url: `${baseUrl}/essay`,
    },
    {
      name: 'By Invitation',
      url: `${baseUrl}/by-invitation`,
    },
    {
      name: 'Schools brief',
      url: `${baseUrl}/schools-brief`,
    },
    {
      name: 'The Economist explains',
      url: `${baseUrl}/the-economist-explains`,
    },
    {
      name: 'The Economist reads',
      url: `${baseUrl}/the-economist-reads`,
    },
  ],
  more: [
    {
      name: 'Newsletters',
      url: `${baseUrl}/newsletters`,
    },
    {
      name: 'Podcasts',
      url: `${baseUrl}/podcasts`,
    },
    {
      name: 'Films',
      url: `${baseUrl}/films`,
    },
    {
      name: 'Subscriber events',
      url: 'https://subscriberevents.economist.com',
    },
    {
      name: 'The Economist app',
      url: `${baseUrl}/get-the-app`,
    },
    {
      name: 'Online courses',
      url: 'https://education.economist.com',
    },
  ],
  account: [
    {
      name: 'Manage my account',
      url: 'https://my.economist.com',
    },
    {
      name: 'Log out',
      url: '',
    },
  ],
});

export const getFooterData = ({
  baseUrl = urls.economist,
  amp = false,
} = {}) => ({
  customerLinks: [
    {
      name: 'Subscribe',
      url: `${baseUrl}/subscribe`,
    },
    {
      name: 'Group subscriptions',
      url: `${baseUrl}/group-subscriptions/`,
    },
    {
      name: 'Reuse our content',
      url: 'https://rights.economist.com/',
    },
    {
      name: 'The Trust Project',
      url: `${baseUrl}/frequently-asked-questions`,
    },
    {
      name: 'Help and contact us',
      url: `https://myaccount.economist.com/s/help`,
    },
  ],
  economistGroupLinks: [
    { name: 'The Economist Group', url: 'https://www.economistgroup.com/' },
    { name: 'Economist Intelligence', url: 'https://www.eiu.com/n/' },
    {
      name: 'Economist Impact',
      url: 'https://impact.economist.com/',
    },
    {
      name: 'Economist Events',
      url:
        'https://events.economist.com/?RefID=e.com-home&utm_source=e.com&utm_medium=website&utm_campaign=group-aff&utm_content=footer-link',
    },
    { name: 'Working Here', url: 'https://www.economistgroup.com/careers' },
    {
      name: 'Economist Education Courses',
      url: 'https://education.economist.com',
    },
    { name: 'Which MBA?', url: 'https://whichmba.economist.com/' },
    { name: 'Executive Jobs', url: 'http://jobs.economist.com/' },
    {
      name: 'Executive Education Navigator',
      url: 'https://execed.economist.com',
    },
  ],
  minorLinks: [
    {
      id: 'footer-terms-of-use',
      name: 'Terms of Use',
      url: 'https://www.economistgroup.com/terms-of-use',
    },
    {
      id: 'footer-privacy',
      name: 'Privacy',
      url: 'https://www.economistgroup.com/privacy-policy',
    },
    {
      id: 'footer-cookie-policy',
      name: 'Cookie Policy',
      url: 'https://www.economistgroup.com/cookies-policy',
    },
    {
      id: 'footer-manage-cookies',
      name: 'Manage Cookies',
      url: '#',
      amp: false,
      className: 'js-cmp-manage-cookies',
    },
    {
      id: 'footer-manage-cookies',
      name: 'Manage Cookies',
      on: 'tap:evidon-amp-consent.prompt()',
      amp: true,
      url: '#footer-manage-cookies',
    },
    {
      id: 'footer-accessibility',
      name: 'Accessibility',
      url: 'https://www.economistgroup.com/accessibility-policy',
    },
    {
      id: 'footer-modern-slavery-statement',
      name: 'Modern Slavery Statement',
      url: 'https://www.economistgroup.com/esg/modern-slavery-act-statement',
    },
    {
      id: 'sitemap',
      name: 'Sitemap',
      url: 'https://www.economist.com/sitemap',
    },
    {
      id: 'footer-ccpa',
      name: 'California: Do Not Sell My Personal Information',
      url: 'https://www.economist.com/privacyportal',
    },
  ].filter(item => typeof item.amp === 'undefined' || item.amp === amp),
  socialLinks: {
    FACEBOOK_LINK: 'https://www.facebook.com/theeconomist',
    INSTAGRAM_LINK: 'https://www.instagram.com/theeconomist',
    TWITTER_LINK: 'https://www.twitter.com/theeconomist',
    LINKEDIN_LINK: 'https://www.linkedin.com/company/the-economist',
    YOUTUBE_LINK: 'https://www.youtube.com/user/economistmagazine',
    RSS_LINK: 'https://www.economist.com/rss',
  },
  theEconomistLinks: [
    {
      name: 'About',
      url: 'https://www.economistgroup.com/about-us',
    },
    { name: 'Advertise', url: 'https://impact.economist.com/' },
    { name: 'Press centre', url: 'https://www.economistgroup.com/group-news' },
  ],
});

export const keyCodes = {
  TAB: 9,
  ESC: 27,
  SPACE: 32,
  ENTER: 13,
};

export const keyValues = {
  TAB: 'Tab',
  ESC: 'Escape',
};

export const reactElementPropType = PropTypes.oneOfType([
  PropTypes.object,
  PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
]);
