// @economist/design-system/src/stories/masthead.stories.js
import React from 'react';
import Masthead from '../common/components/masthead/masthead';
import IconCheck from '../common/components/assets/icon/icon-check';

export default {
  title: 'common/Masthead',
  component: Masthead,
  parameters: {
    layout: 'fullscreen',
  },
  args: {
    subscribeCtaText: 'Subscribe',
    logoutUrl: '/log-out',
    manageMyAccountUrl: '/my/account',
    savedStoriesUrl: '/saved/stories',
    gsClientAdminUrl: '/client-admin',
    mySubscriptionUrl: '/my-subscription',
    personalInfoUrl: '/personal-info',
    billingUrl: '/billing',
    newslettersAndPreferencesUrl: '/newsletters-and-preferences',
    giftSubscriptionsUrl: '/gift-subscriptions',
    helpUrl: '/help',
    isClientAdmin: false,
    isGroupSubscriber: false,
    hasTemporaryExtraItem: false,
    fullName: 'John Smith',
    userVariant: 'anonymous',
  },
  argTypes: {
    userVariant: {
      control: {
        type: 'select',
        options: ['anonymous', 'registered', 'subscriber'],
      },
    },
  },
};

const url = 'https://example.com/auth';

export const masthead = args => {
  const { hasTemporaryExtraItem, userVariant, ...props } = args;
  const tei = hasTemporaryExtraItem && (
    <>
      <IconCheck />
    </>
  );
  switch (userVariant) {
    case 'anonymous':
      return (
        <Masthead
          {...props}
          temporaryExtraItem={tei}
          loginUrl={url}
          showSubscribe
          isLoggedIn={false}
        />
      );
    case 'registered':
      return (
        <Masthead
          {...props}
          temporaryExtraItem={tei}
          showSubscribe
          isLoggedIn
        />
      );
    case 'subscriber':
      return (
        <Masthead
          {...props}
          temporaryExtraItem={tei}
          showSubscribe={false}
          isLoggedIn
        />
      );
    default:
      return <>userVariant not recognized</>;
  }
};
