// @economist/design-system/src/common/components/media-promo/media-promo.js
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import 'reset-css/reset.css';
import '../color/color';
import '../grid/grid';
import '../grid/layout-grid';
import '../typography/typography';
import css from './media-promo.css';
import LinkWithArrowIcon from '../link-with-arrow-icon/link-with-arrow-icon';

const MediaPromo = ({
  title,
  description,
  inverse,
  variant,
  linkText,
  linkURL,
}) => (
  <div
    className={classnames(css['ds-media-promo'], {
      [css['ds-media-promo--inverse']]: inverse,
      [css['ds-media-promo--podcasts']]: variant === 'podcasts',
      [css['ds-media-promo--films']]: variant === 'films',
      [css['ds-media-promo--editorial']]: variant === 'editorial',
    })}
  >
    <div className={css['ds-layout-media-promo']}>
      <div className={`ds-layout-grid ${css['ds-media-promo__group']}`}>
        <h2 className={css['ds-media-promo__title']}>{title}</h2>
        <p className={css['ds-media-promo__description']}>{description}</p>
        <LinkWithArrowIcon
          href={linkURL}
          text={linkText}
          inverse={inverse}
          heavy
          minor
        />
      </div>
    </div>
  </div>
);

MediaPromo.defaultProps = {
  description: '',
  inverse: false,
};

MediaPromo.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  linkText: PropTypes.string.isRequired,
  linkURL: PropTypes.string.isRequired,
  variant: PropTypes.oneOf(['films', 'podcasts', 'editorial']).isRequired,
  inverse: PropTypes.bool,
};

export default MediaPromo;
