// @economist/design-system/src/common/components/banner-alert/banner-alert.js
import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import 'reset-css/reset.css';
import '../color/color';
import '../typography/typography';
import ControlClose from '../controls/control-close';
import css from './banner-alert.css';

const BannerAlert = ({
  button,
  closable,
  description,
  handleClose,
  isClosed,
  text,
  variant,
  ...otherProps
}) => {
  if (isClosed) {
    return null;
  }
  return (
    <div
      className={classnames(css['ds-banner-alert'], {
        [css[`ds-banner-alert--${variant}`]]: variant,
      })}
      {...otherProps}
    >
      <div className={css['ds-banner-alert-inner']}>
        <p className={css['ds-banner-alert__text']}>
          <strong>{text}</strong> {description}
        </p>
        {button && (
          <div className={css['ds-banner-alert-button']}>{button}</div>
        )}
      </div>
      {closable && <ControlClose onClick={handleClose} />}
    </div>
  );
};

BannerAlert.defaultProps = {
  button: null,
  closable: true,
  description: '',
  handleClose: null,
  isClosed: false,
  text: '',
  variant: null,
};

BannerAlert.propTypes = {
  button: PropTypes.node,
  closable: PropTypes.bool,
  description: PropTypes.string,
  handleClose: PropTypes.func,
  isClosed: PropTypes.bool,
  text: PropTypes.string,
  variant: PropTypes.oneOf(['error', 'success', 'warning']),
};

export default BannerAlert;
